Support DM9051 SPI ethernet device

bmork opened the pull request:
# What does this implement/fix?

Adds support for the Davicom DM9051 SPI Ethernet Controller

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes https://github.com/esphome/feature-requests/issues/2427

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#3914

## Test Environment

- [x] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx

## Example entry for `config.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example config.yaml for the ETH01-Evo board
ethernet:
  type: DM9051
  clk_pin: GPIO07
  mosi_pin: GPIO10
  miso_pin: GPIO03
  cs_pin: GPIO09
  interrupt_pin: GPIO08
  reset_pin: GPIO06
  clock_speed: 8MHz
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

bmork committed:
[ethernet] Add DM9051 support

Fixes https://github.com/esphome/feature-requests/issues/2427

Tested on an ETH01-Evo board using

  ethernet:
    type: DM9051
    clk_pin: GPIO07
    mosi_pin: GPIO10
    miso_pin: GPIO03
    cs_pin: GPIO09
    interrupt_pin: GPIO08
    reset_pin: GPIO06
    clock_speed: 20MHz

A number of read errors was observed with 26 MHz
(default) or higher clock speeds. This might be a
board specific problem.

Signed-off-by: Bjørn Mork <bjorn@mork.no>

bmork committed:
[ethernet] prepare for more SPI devices

The esp-idf SPI ethernet drivers will set command_bits
and address_bits if the fields are 0, allowing this
config to be shared.

Signed-off-by: Bjørn Mork <bjorn@mork.no>

JakobTewes commented:
Heyhey,

got one of the "DM9051 SPI ethernet" equipped ESP32-C3 (eth01-evo) and am willing to review. 
Unfortunately, I have now idea what to do and how to build. Can someone please elaborate?

bmork commented:
@JakobTewes I'm no expert here, but luckily this question is answered in
https://esphome.io/guides/faq.html#how-can-i-test-a-pull-request

So just use your reasonably uptodate esphome installation and add an external_component reference to this pull request. Something like
```yaml
esphome:
  name: eth01-evo

esp32:
  variant: esp32c3
  board: esp32-c3-devkitm-1
  framework:
    type: esp-idf
    version: latest
    platform_version: 6.5.0

external_components:
  - source: github://pr#6861
    components:
      - ethernet

logger:
  hardware_uart: UART0

api:
ota:
  platform: esphome

ethernet:
  type: DM9051
  clk_pin: GPIO07
  mosi_pin: GPIO10
  miso_pin: GPIO03
  cs_pin: GPIO09
  interrupt_pin: GPIO08
  reset_pin: GPIO06
  clock_speed: 20MHz
```
And anything else you like of course.  Then simply build and load the firmware like usual. E.g. `esphome run test.yaml`

jesserockz commented on esphome/components/ethernet/__init__.py:
```diff
@@ -218,6 +220,7 @@ async def to_code(config):
         if CORE.using_esp_idf:
             add_idf_sdkconfig_option("CONFIG_ETH_USE_SPI_ETHERNET", True)
             add_idf_sdkconfig_option("CONFIG_ETH_SPI_ETHERNET_W5500", True)
+            add_idf_sdkconfig_option("CONFIG_ETH_SPI_ETHERNET_DM9051", True)
```
We should be only setting the one that the user is actually using here

jesserockz commented on esphome/components/ethernet/ethernet_component.cpp:
```diff
@@ -107,19 +107,22 @@ void EthernetComponent::setup() {
 
 #if USE_ESP_IDF && (ESP_IDF_VERSION_MAJOR >= 5)
   eth_w5500_config_t w5500_config = ETH_W5500_DEFAULT_CONFIG(host, &devcfg);
+  eth_dm9051_config_t dm9051_config = ETH_DM9051_DEFAULT_CONFIG(host, &devcfg);
```
This code will need restructuring to handle different spi chipsets.
As above, we shouldn't be creating both (more if more are added)

bmork replied:
I was a bit unsure about that.  Thanks.

The code becomes butt ugly with ifdefs due to the combination of per-device config objects and different driver init between esp-idf v4 and v5.  But I guess that's how it must be at the moment.  We can clean this up when the default esp-idf version is updated.

bmork commented:
AFAIU the current SPI ethernet implementation depends on esp-idf, so I wonder if it is appropriate to raise an error in the python code if a SPI device is selected using another framework?

This seems more user friendly than failing silently to me.  And it simplifies all the ifdeffery somewhat.  But it also means that we have to delete test-w5500.esp32-ard.yaml since it will fail to build (intentionally).

JakobTewes commented:
Built and installed successfully with these options:

`esphome:
  name: $devicename
  comment: ${device_description}
  build_path: esp32
  platformio_options:
    upload_speed: 115200

esp32:
  board: esp32-c3-devkitm-1
# currently needed, but not in the future
  variant: esp32c3
  framework:
    type: esp-idf
    version: latest
    platform_version: 6.5.0

external_components:
  - source: github://pr#6861
    components:
      - ethernet

ethernet:
# https://github.com/esphome/esphome/pull/6861
  type: DM9051
  clk_pin: GPIO07
  mosi_pin: GPIO10
  miso_pin: GPIO03
  cs_pin: GPIO09
  interrupt_pin: GPIO08
  reset_pin: GPIO06
  clock_speed: 8MHz`

JakobTewes commented:
Now also tested via wireless upload and works like a charm.

toxic0berliner commented:
hmm.. 🤔 Somehow I'm unable to build this...
This is what I get
<details><summary>build logs</summary>

```
INFO ESPHome 2024.8.3
INFO Reading configuration /config/esphome/eth01.yaml...
WARNING The selected ESP-IDF framework version is not the recommended one. If there are connectivity or build issues please remove the manual version.
WARNING The selected ESP-IDF framework version is not the recommended one. If there are connectivity or build issues please remove the manual version.
WARNING GPIO9 is a strapping PIN and should only be used for I/O with care.
Attaching external pullup/down resistors to strapping pins can cause unexpected failures.
See https://esphome.io/guides/faq.html#why-am-i-getting-a-warning-about-strapping-pins
WARNING GPIO8 is a strapping PIN and should only be used for I/O with care.
Attaching external pullup/down resistors to strapping pins can cause unexpected failures.
See https://esphome.io/guides/faq.html#why-am-i-getting-a-warning-about-strapping-pins
INFO Generating C++ source...
INFO Updating https://github.com/espressif/esp-protocols.git@mdns-v1.2.5
INFO Compiling app...
Processing eth01-evo (board: esp32-c3-devkitm-1; framework: espidf; platform: platformio/espressif32@6.5.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32C3 160MHz, 320KB RAM, 4MB Flash
 - framework-espidf @ 3.50102.240122 (5.1.2) 
 - tool-cmake @ 3.16.4 
 - tool-ninja @ 1.7.1 
 - tool-riscv32-esp-elf-gdb @ 12.1.0+20221002 
 - tool-xtensa-esp-elf-gdb @ 12.1.0+20221002 
 - toolchain-esp32ulp @ 2.35.0-20220830 
 - toolchain-riscv32-esp @ 12.2.0+20230208
Reading CMake configuration...
-- git rev-parse returned 'fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).'
-- The C compiler identification is unknown
-- The CXX compiler identification is unknown
-- The ASM compiler identification is GNU
-- Found assembler: /data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc
-- Check for working C compiler: /data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc
-- Check for working C compiler: /data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc -- broken
-- Configuring incomplete, errors occurred!
See also "/data/esp32/.pioenvs/eth01-evo/CMakeFiles/CMakeOutput.log".
See also "/data/esp32/.pioenvs/eth01-evo/CMakeFiles/CMakeError.log".

fatal: not a git repository (or any parent up to mount point /)
Stopping at filesystem boundary (GIT_DISCOVERY_ACROSS_FILESYSTEM not set).
CMake Error at /data/cache/platformio/packages/tool-cmake/share/cmake-3.16/Modules/CMakeTestCCompiler.cmake:60 (message):
  The C compiler

    "/data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /data/esp32/.pioenvs/eth01-evo/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/data/cache/platformio/packages/tool-ninja/ninja cmTC_8fadc && [1/2] Building C object CMakeFiles/cmTC_8fadc.dir/testCCompiler.c.obj
    FAILED: CMakeFiles/cmTC_8fadc.dir/testCCompiler.c.obj 
    /data/cache/platformio/packages/toolchain-riscv32-esp/bin/riscv32-esp-elf-gcc   -march=rv32imc_zicsr_zifencei -o CMakeFiles/cmTC_8fadc.dir/testCCompiler.c.obj   -c testCCompiler.c
    riscv32-esp-elf-gcc: fatal error: cannot execute 'as': execvp: No such file or directory
    compilation terminated.
    ninja: build stopped: subcommand failed.
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  /data/cache/platformio/packages/framework-espidf/tools/cmake/project.cmake:448 (__project)
  CMakeLists.txt:3 (project)



========================== [FAILED] Took 1.99 seconds ==========================
```
</details>

using the following config
<details><summary>config</summary>

```
esphome:
  name: eth01-evo
  friendly_name: ETH01-evo
  build_path: esp32
  platformio_options:
    upload_speed: 115200

esp32:
  variant: esp32c3
  board: esp32-c3-devkitm-1
  framework:
    type: esp-idf
    version: latest
    platform_version: 6.5.0

external_components:
  - source: github://pr#6861
    components:
      - ethernet

ethernet:
# https://github.com/esphome/esphome/pull/6861
  type: DM9051
  clk_pin: GPIO07
  mosi_pin: GPIO10
  miso_pin: GPIO03
  cs_pin: GPIO09
  interrupt_pin: GPIO08
  reset_pin: GPIO06
  clock_speed: 8MHz
```

Any idea ?

bmork commented:
rebased onto dev after the openeth PR (#7020) was merged

bmork commented:
force pushed resolved conflicts after 302ba2874e12 ("Support W5500 SPI-Ethernet polling mode if framework is supported (#7503)")

and added polling mode support for DM9051 too using that  implementation.

bmork commented:
@jesserockz ping?

bmork commented:
> It is no longer possible to use your component for the `DM9051` controller.

This is because it now depends on https://github.com/esphome/esphome/commit/302ba2874e12960c76a624fa6066bf5df1ffac2e , which is only in the dev branch at the moment.  The PR has to apply to dev to be mergeable.

You can still use the previous version of the PR as an external component.  It's a little harder to configure since you have to point to https://github.com/esphome/esphome/commit/d5bdf2575c151abfad39b6c24c3856395a7d0901 instead of the PR shorthand version.

Or you can wait for the next esphome release, where https://github.com/esphome/esphome/commit/302ba2874e12960c76a624fa6066bf5df1ffac2e will be included at least.  The the PR shorthand should work again.  Until I have to fix some otther conflict with new code in the ethernet component...

> When will your changes be merged?

You tell me.  I have no idea

Sleeper85 commented:
> You can still use the previous version of the PR as an external component. It's a little harder to configure since you have to point to [d5bdf25](https://github.com/esphome/esphome/commit/d5bdf2575c151abfad39b6c24c3856395a7d0901) instead of the PR shorthand version.

How do it?

What would be the address of the source component?

bmork commented:
> What would be the address of the source component?

Something like
```yaml
external_components:
  - source: 
      type: git
      url: https://github.com/esphome/esphome
      ref: d5bdf2575c151abfad39b6c24c3856395a7d0901
    components:
      - ethernet
```

Gojjin commented:
Any one can help i am using the board eth01-evo with poe . i am finding it hard to find how to configure the ethernet port to be able to connect over my wired network. i am using Aruduino IDE as a compiler, anyone have an example please to follow to make this work ? 
Thanks

klim123123 commented:
> Any one can help i am using the board eth01-evo with poe . i am finding it hard to find how to configure the ethernet port to be able to connect over my wired network. i am using Aruduino IDE as a compiler, anyone have an example please to follow to make this work ? Thanks

works for me:
```
esphome:
  name: eth01-evo
  friendly_name: eth01-evo

esp32:
  board: esp32-c3-devkitm-1
# currently needed, but not in the future
  variant: esp32c3
  framework:
    type: esp-idf
    version: latest
    platform_version: 6.5.0

external_components:
  - source: github://pr#6861
    components:
      - ethernet

ethernet:
# https://github.com/esphome/esphome/pull/6861
  type: DM9051
  clk_pin: GPIO07
  mosi_pin: GPIO10
  miso_pin: GPIO03
  cs_pin: GPIO09
  interrupt_pin: GPIO08
  reset_pin: GPIO06
  clock_speed: 20MHz
```

Gojjin commented:
> > Any one can help i am using the board eth01-evo with poe . i am finding it hard to find how to configure the ethernet port to be able to connect over my wired network. i am using Aruduino IDE as a compiler, anyone have an example please to follow to make this work ? Thanks
> 
> works for me:
> 
> ```
> esphome:
>   name: eth01-evo
>   friendly_name: eth01-evo
> 
> esp32:
>   board: esp32-c3-devkitm-1
> # currently needed, but not in the future
>   variant: esp32c3
>   framework:
>     type: esp-idf
>     version: latest
>     platform_version: 6.5.0
> 
> external_components:
>   - source: github://pr#6861
>     components:
>       - ethernet
> 
> ethernet:
> # https://github.com/esphome/esphome/pull/6861
>   type: DM9051
>   clk_pin: GPIO07
>   mosi_pin: GPIO10
>   miso_pin: GPIO03
>   cs_pin: GPIO09
>   interrupt_pin: GPIO08
>   reset_pin: GPIO06
>   clock_speed: 20MHz
> ```

Thanks for your reply, however I am note using esp home as I have a simple project to get temperature and humidity reading and sending the data via SNMP oid so the data can be read by a software PRTG.  i went for this board because of the POE capabilities. I have another board with ESP32 Wroom and the ethernet configuration is working fine (but no poe capabilities) so the board ETH01-EVO POE with esp32-C3 is giving me a challenge which libraries should work or not maybe anyone had the same issue ?

Imanol82 commented:
Sorry for reopening the topic but why if I update from 2024.11.3 to 2024.12.2 do the ESP32 ETH01-EVO POE stop working?

JakobTewes commented:
Can confirm - this works with 2024.12.2:

`esp32:
  board: esp32-c3-devkitm-1
# currently needed, but not in the future
  variant: esp32c3
  framework:
    type: esp-idf
    version: latest
#    platform_version: 6.5.0

external_components:
  - source: github://pr#6861
    components:
      - ethernet

ethernet:
# https://github.com/esphome/esphome/pull/6861
  type: DM9051
  clk_pin: GPIO07
  mosi_pin: GPIO10
  miso_pin: GPIO03
  cs_pin: GPIO09
  interrupt_pin: GPIO08
  reset_pin: GPIO06
  clock_speed: 8MHz`

Imanol82 commented:
Hello. 

I've received this error when try to update the esp module:

![image](https://github.com/user-attachments/assets/b67f661d-f041-43d7-9b16-89f370520036)

```
substitutions:
  settings_skipped_updates: "30"
  devicename: "baxi_esp"

esphome:
  name: $devicename

esp32:
  board: esp32-c3-devkitm-1
  variant: esp32c3
  framework:
    type: esp-idf
    version: latest
    platform_version: 6.5.0

external_components:
  - source: 
      type: git
      url: https://github.com/esphome/esphome
      ref: d5bdf2575c151abfad39b6c24c3856395a7d0901
    components:
      - ethernet
```

I have commented version line and that's all.

```
#    platform_version: 6.5.0
```

Thank you very much!!

giovi321 commented:
Hello, is there any news on the release of the support for those WT32-ETH01 like devices?
Many thanks!

sabaatworld commented:
Would love to see this merged in. I can confirm that https://github.com/esphome/esphome/pull/6861#issuecomment-2571396522 works well.

giovi321 commented:
Update from my side: tested with a WT32-ETH01-EVO-PoE, works like a charm. Here's the code that somebody else already mentioned.

```
esphome:
  name: eth01-evo
  friendly_name: eth01-evo

esp32:
  board: esp32-c3-devkitm-1
  variant: esp32c3
  framework:
    type: esp-idf
    version: latest
 
external_components:
  - source: github://pr#6861
    components:
      - ethernet
 
ethernet:
# https://github.com/esphome/esphome/pull/6861
  type: DM9051
  clk_pin: GPIO07
  mosi_pin: GPIO10
  miso_pin: GPIO03
  cs_pin: GPIO09
  interrupt_pin: GPIO08
  reset_pin: GPIO06
  clock_speed: 8MHz

logger:

api:

ota:
  platform: esphome
```

tht commented:
Same issue as @Imanol82 while trying to update to 2025.2.0. It worked perfectly fine before.
```
AttributeError: module 'esphome.config_validation' has no attribute 'ipv4'
Failed config
```
I assume something changed with the new version and this PR needs to be updated.

randybb commented:
There was a change https://github.com/esphome/esphome/pull/8072 and since the whole ethernet component is pulled from the downloaded external component and it has been not updated, so it is not compatible with the current version. I guess cleaning build files OR refreshing component by adding `refresh: 0s` should help.

tht commented:
> I rebased onto the current "dev" branch. Completely untested since I've pretty much given up on this PR.. But maybe it works?

You're my hero :). It works perfectly fine first try. Thank you so much. I hope this will get merged in the future.

Imanol82 commented:
Hello!

I have updated again and today it works. Thank you very much!

giovi321 commented:
Can't build anymore... I get the same error as @Imanol82. Anybody managed to build in the last 4 days?

jesserockz commented on esphome/components/ethernet/__init__.py:
```diff
@@ -233,6 +235,8 @@ def _validate(config):
 def _final_validate(config):
     if config[CONF_TYPE] not in SPI_ETHERNET_TYPES:
         return
+    if not CORE.using_esp_idf:
+        raise cv.Invalid("SPI ethernet requires ESP-IDF")
```
Why is this a requirement?

bmork replied:
Short circuit on my side.  I didn't realize these drivers were built into the arduino framework by default.  Sorry about that, and thanks a lot for the thorough review.

Have dropped this patch from the series, added back the multi-level ifdef'ery in  ethernet_component.cpp and added an arduino test.

fnord123 commented:
This worked fine for me on the dev branch, except I had to set the clock_speed for the ethernet entry to be 20Mhz rather than 8Mhz for it to work correctly.

Imanol82 commented:
Hello!!

It has broken again? 

![image](https://github.com/user-attachments/assets/e0f63991-ad08-460f-ab8f-2cedf58346da)

bmork commented:
> It has broken again?

I can't reproduce this.  Your config (excluding the static IP) works for me on esphome 2025.4.1

Imanol82 commented:
I don't know what happened. One day all ESPs fails and next day all worked.

No idea, sorry for the message.

jesserockz commented on esphome/components/ethernet/__init__.py:
```diff
@@ -295,7 +297,12 @@ async def to_code(config):
         if CORE.using_esp_idf:
             add_idf_sdkconfig_option("CONFIG_ETH_USE_SPI_ETHERNET", True)
-            add_idf_sdkconfig_option("CONFIG_ETH_SPI_ETHERNET_W5500", True)
+            add_idf_sdkconfig_option(
+                "CONFIG_ETH_SPI_ETHERNET_W5500", config[CONF_TYPE] == "W5500"
+            )
+            add_idf_sdkconfig_option(
+                "CONFIG_ETH_SPI_ETHERNET_DM9051", config[CONF_TYPE] == "DM9051"
+            )
```
```suggestion
            add_idf_sdkconfig_option(
                f"CONFIG_ETH_SPI_ETHERNET_{config[CONF_TYPE]}", True
            )
```

jesserockz reviewed (changes requested):
Small change below, otherwise good to go.

jesserockz committed:
Merge branch 'dev' into dm9051

bmork committed:
[ethernet] simplify sdkconfig for SPI devices

Simplifies the code and makes it easier to add more ethernet SPI devices in the future

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

Imanol82 commented:
Hello! The same happens again with the new ESPHome:

![image](https://github.com/user-attachments/assets/c8509b4a-51f0-4308-b817-86a28f5a5d28)

This is my config, I think is the same as always...:

![image](https://github.com/user-attachments/assets/171be7df-fd43-4358-a0da-6665f5600431)

bmork committed:
Merge branch 'dev' into dm9051

Wiktor173 commented:
Hi, when trying to use with ESPHome 2025.6.1, I get the error:

src/esphome/components/ethernet/ethernet_component.cpp: In member function 'virtual void esphome::ethernet::EthernetComponent::loop()':
src/esphome/components/ethernet/ethernet_component.cpp:310:15: error: 'class esphome::ethernet::EthernetComponent' has no member named 'disable_loop'
  310 |         this->disable_loop();
      |               ^~~~~~~~~~~~
src/esphome/components/ethernet/ethernet_component.cpp: In static member function 'static void esphome::ethernet::EthernetComponent::eth_event_handler(void*, esp_event_base_t, int32_t, void*)':
src/esphome/components/ethernet/ethernet_component.cpp:438:29: error: 'class esphome::ethernet::EthernetComponent' has no member named 'enable_loop_soon_any_context'
  438 |       global_eth_component->enable_loop_soon_any_context();
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/ethernet/ethernet_component.cpp:444:29: error: 'class esphome::ethernet::EthernetComponent' has no member named 'enable_loop_soon_any_context'
  444 |       global_eth_component->enable_loop_soon_any_context();  // Enable loop when connection state changes
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/ethernet/ethernet_component.cpp:452:29: error: 'class esphome::ethernet::EthernetComponent' has no member named 'enable_loop_soon_any_context'
  452 |       global_eth_component->enable_loop_soon_any_context();  // Enable loop when connection state changes
      |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/esphome/components/ethernet/ethernet_component.cpp: In static member function 'static void esphome::ethernet::EthernetComponent::got_ip_event_handler(void*, esp_event_base_t, int32_t, void*)':
src/esphome/components/ethernet/ethernet_component.cpp:472:25: error: 'class esphome::ethernet::EthernetComponent' has no member named 'enable_loop_soon_any_context'
  472 |   global_eth_component->enable_loop_soon_any_context();  // Enable loop when connection state changes
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~

bmork commented:
> when trying to use with ESPHome 2025.6.1, I get the error

This is because the PR is tracking the dev branch, and I had to do  a merge to resolve a conflict in esphome/components/ethernet/ethernet_component.cpp

You can either point your config to the commit before this merge, or build with the "dev" branch, or wait until 2025.7 is released.

Sorry about this.  The real fix is of course to get this PR merged.  AFAIK there are no remaining issues, so this is simply a matter of review resources. Wiich of course isn't so simple.  But let's cross our fingers for the next release.

Wiktor173 commented:
Thank you for your work.
